crates/bevy_ecs/src/query/iter.rs

```diff
@@ -131,52 +131,55 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIter<'w, 's, D, F> {
         &mut self,
         mut accum: B,
         func: &mut Func,
-        storage: StorageId,
+        storage_id: StorageId,
         range: Option<Range<usize>>,
     ) -> B
     where
         Func: FnMut(B, D::Item<'w>) -> B,
     {
-        if self.cursor.is_dense {
-            // SAFETY: `self.cursor.is_dense` is true, so storage ids are guaranteed to be table ids.
-            let table_id = unsafe { storage.table_id };
-            // SAFETY: Matched table IDs are guaranteed to still exist.
-            let table = unsafe { self.tables.get(table_id).debug_checked_unwrap() };
-
-            let range = range.unwrap_or(0..table.entity_count());
-            accum =
-                // SAFETY:
-                // - The fetched table matches both D and F
-                // - caller ensures `range` is within `[0, table.entity_count)`
-                // - The if block ensures that the query iteration is dense
-                unsafe { self.fold_over_table_range(accum, func, table, range) };
-        } else {
-            // SAFETY: `self.cursor.is_dense` is false, so storage ids are guaranteed to be archetype ids.
-            let archetype_id = unsafe { storage.archetype_id };
-            // SAFETY: Matched archetype IDs are guaranteed to still exist.
-            let archetype = unsafe { self.archetypes.get(archetype_id).debug_checked_unwrap() };
-            // SAFETY: Matched table IDs are guaranteed to still exist.
-            let table = unsafe { self.tables.get(archetype.table_id()).debug_checked_unwrap() };
-
-            let range = range.unwrap_or(0..archetype.len());
-
-            // When an archetype and its table have equal entity counts, dense iteration can be safely used.
-            // this leverages cache locality to optimize performance.
-            if table.entity_count() == archetype.len() {
-                accum =
-                // SAFETY:
-                // - The fetched archetype matches both D and F
-                // - The provided archetype and its' table have the same length.
-                // - caller ensures `range` is within `[0, archetype.len)`
-                // - The if block ensures that the query iteration is not dense.
-                unsafe { self.fold_over_dense_archetype_range(accum, func, archetype,range) };
-            } else {
+        match self.cursor.entity_source {
```

Comment:
I considered that, but wasn't sure if this would just invert the unsafety - because now the caller passing the incorrect type would cause you to perform the wrong kind of iteration, which may have bizarre side-effects. There could also separately be a debug assert that just checks the storage ID variant matches the cursor's iteration source?

What do you think is best?

Reply:
I'd vote for changing it.  They need to be the same type no matter which way you do the `match`, but at least the other way has one fewer `unsafe` thing to check.  And I think it lets you get rid of the `debug_checked_as_` methods entirely.  

We probably should note in the safety requirements for the method that the type needs to match.  But the safety requirements were already incomplete: The method body asserted that "Matched table IDs are guaranteed to still exist", but the safety comments didn't actually require the caller to pass any particular value for `storage`.  So maybe the requirement should be that `storage_id` was obtained from `entity_source`?  

Actually... it might even be sound to pass the wrong kind of storage, as long as all entities in the storage match the query!

Reply:
You would, I think, always get away with iterating archetypes when you could've been doing a dense table iteration.

The other way, iterating tables when it's expecting you to iterate archetypes, is fine as long as the caller asks you to visit sensible subsets of tables. If it just blindly iterated tables instead of archetypes, you'd get issues around interleaving entities from multiple archetypes, and skip entities with only sparse set components and no table components.

Either way: you're right that flipping this will remove the need for the unsafe accessors on `StorageId` though, and as you say I can just comment that the method should only provide a `StorageId` that has come from the `entity_source` to describe what I believe are the same requirements.